src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -58,46 +58,107 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
                 // and the generic won't be "partially written".
-                if (testPosition == position)
+                if (spanStart == position)
                 {
-                    var typeArgumentList = leftToken.GetAncestor<TypeArgumentListSyntax>();
-                    if (typeArgumentList != null)
-                    {
-                        if (typeArgumentList.LessThanToken != default && typeArgumentList.GreaterThanToken != default)
-                        {
-                            testPosition = typeArgumentList.LessThanToken.SpanStart;
-                        }
-                    }
+                    spanStart = token.GetAncestor<GenericNameSyntax>()?.SpanStart ?? spanStart;
+                }
+
+                var tokenLeftOfGenericName = syntaxTree.FindTokenOnLeftOfPosition(spanStart, cancellationToken);
+                if (tokenLeftOfGenericName.IsKind(SyntaxKind.DotToken) && tokenLeftOfGenericName.Parent.IsKind(SyntaxKind.QualifiedName))
+                {
+                    spanStart = tokenLeftOfGenericName.Parent.SpanStart;
                 }
             }
 
-            if ((!leftToken.GetPreviousTokenIfTouchingWord(position).IsKindOrHasMatchingText(SyntaxKind.AsyncKeyword) &&
-                syntaxTree.IsMemberDeclarationContext(testPosition, contextOpt: null, validModifiers: SyntaxKindSet.AllMemberModifiers, validTypeDeclarations: SyntaxKindSet.ClassInterfaceStructTypeDeclarations, canBePartial: false, cancellationToken: cancellationToken)) ||
-                syntaxTree.IsStatementContext(testPosition, leftToken, cancellationToken) ||
-                syntaxTree.IsGlobalMemberDeclarationContext(testPosition, SyntaxKindSet.AllGlobalMemberModifiers, cancellationToken) ||
-                syntaxTree.IsGlobalStatementContext(testPosition, cancellationToken) ||
-                syntaxTree.IsDelegateReturnTypeContext(testPosition, syntaxTree.FindTokenOnLeftOfPosition(testPosition, cancellationToken), cancellationToken))
+            return spanStart;
+        }
+
+        private static int WalkOutOfRefType(SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        {
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var prevToken = token.GetPreviousTokenIfTouchingWord(position);
+
+            if (prevToken.IsKind(SyntaxKind.RefKeyword, SyntaxKind.ReadOnlyKeyword) && prevToken.Parent.IsKind(SyntaxKind.RefType))
             {
-                return true;
+                return prevToken.SpanStart;
             }
 
-            return false;
+            return position;
+        }
+
+        private static int WalkOutOfTupleType(SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        {
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var prevToken = token.GetPreviousTokenIfTouchingWord(position);
```

Comment:
Nitpick: you might just want to write this syntaxTree.FindTokenOnLeftOfPosition(…).GetPreviousTokenIfTouchingWord() so then you don't have two locals and can't accidentally use the wrong one.

Reply:
Sure.